File: c++/c19_2_hamiltonian_path.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;
using ull = unsigned long long;

namespace {
	bool ore(const vector<vector<ll>>& adj) {
		ll n = ssize(adj);
		auto adj2 = adj | views::transform(ranges::to<unordered_set>()) | ranges::to<vector>();
		for (ll i = 0; i < n - 1; ++i) {
			for (ll j = i + 1; j < n; ++j) {
				if (!adj2[i].contains(j) && ssize(adj[i]) + ssize(adj[j]) < n) {
					return false;
				}
			}
		}
		return true;
	}

	vector<ll> construction(const vector<vector<ll>>& adj) {
		ll n = ssize(adj);
		vector<ll> adj2(adj.size());
		for (ll i = 0; i < n; ++i) {
			for (ll j : adj[i]) {
				adj2[i] |= 1ll << j;
			}
		}
		vector<ll> dp(1ll << n);
		for (ll s = 1; s < 1ll << n; ++s) {
			if ((s & (s - 1)) == 0)
				dp[s] = s;
			else {
				for (ll i = s; i != 0; i &= i - 1) {
					ll j = i & -i;
					if (dp[s ^ j] & adj2[bit_width(ull(j)) - 1])
						dp[s] |= j;
				}
			}
		}
		ll s = (1ll << n) - 1, prev = s;
		vector<ll> ret;
		while (s) {
			ll c = prev & dp[s], i = c & -c, j = bit_width(ull(i)) - 1;
			ret.push_back(j);
			prev = adj2[j];
			s ^= i;
		}
		return ret;
	}

	void test(const vector<vector<ll>>& adj) {
		auto ret = construction(adj);
		for (auto [i, j] : ret | views::pairwise)
			EXPECT_TRUE(ranges::find(adj[i], j) != adj[i].end());
		ranges::sort(ret);
		for (ll i = 0; i < ssize(ret); ++i)
			EXPECT_EQ(ret[i], i);
	}
}

TEST(C192HamiltonianPath, ore) {
	EXPECT_TRUE(ore({}));
	EXPECT_FALSE(ore({ {1,3},{0,2,4},{1,4},{0,4},{1,2,3} }));
}

TEST(C192HamiltonianPath, construction) {
	test({});
	test({ {1,3},{0,2,4},{1,4},{0,4},{1,2,3} });
}

File: python/c19_2_hamiltonian_path.py
import unittest
from itertools import combinations,pairwise

def ore(adj):
    n=len(adj)
    adj=[set(e) for e in adj]
    return all(len(adj[i])+len(adj[j])>=n-1 for i,j in combinations(range(1,n),2) if j not in adj[i])

def construction(adj):
    n=len(adj)-1
    adj=[sum(1<<(j-1) for j in adj[i]) for i in range(1,n+1)]
    dp=[0]*(1<<n)
    for s in range(1,1<<n):
        if s.bit_count()<=1:
            dp[s]=s
        else:
            for i in range(n):
                if s&1<<i and dp[s^1<<i]&adj[i]:
                    dp[s]|=1<<i
    s=(1<<n)-1
    prev=s
    while s:
        for i in range(n):
            if prev&dp[s]&1<<i:
                break
        yield i+1
        prev=adj[i]
        s^=1<<i

class Test(unittest.TestCase):
    def assertConstruction(self,adj):
        with self.subTest(adj=adj):
            ret=[*construction(adj)]
            self.assertCountEqual(ret,range(1,len(adj)))
            for i,j in pairwise(ret):
                self.assertIn(j,adj[i])
    def test(self):
        adj=[[],[2,4],[1,3,5],[2,5],[1,5],[2,3,4]]
        self.assertTrue(ore([[]]))
        self.assertFalse(ore(adj))

        self.assertConstruction([[]])
        self.assertConstruction(adj)

if __name__=='__main__':
    unittest.main()
